Functional workflow for LocalRuntime APIs

payalcha opened the pull request:
Adding functional tests of Workflow APIs into pull request.
Converted all test scripts except datastore_cli to functional test.

payalcha committed:
WorkFlow functional tests

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Functional workflow

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Workflow all functional tests

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Merge branch 'develop' into functional-workflow

payalcha committed:
Functional test

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Merge branch 'functional-workflow' of https://github.com/payalcha/openfl into functional-workflow

payalcha committed:
Functional test

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Functional test

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
fix python installs

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
fix python installs

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
fix python installs

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
keep import conditional

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

rahulga1 commented on .github/workflows/wf_functional_e2e.yml:
```diff
@@ -0,0 +1,95 @@
+          python -m pip install --upgrade pip
+          pip install .
+          pip install -r test-requirements.txt
+          pip install -r openfl-tutorials/experimental/workflow/workflow_interface_requirements.txt
```
you might have to ensure there is no conflict in terms of package version between these two files.

payalcha replied:
Removed one of the requirements.txt

noopurintel commented on tests/github/experimental/workflow/LocalRuntime/testflow_subset_of_collaborators.py:
```diff
@@ -142,6 +144,7 @@ def callable_to_initialize_collaborator_private_attributes(collab_name):
         collaborators_ran = testflow_subset_collaborators.collaborators_ran
         # We now convert names to lowercase
         random_ints = testflow_subset_collaborators.random_ints
+        print(f"random_ints = {random_ints}")
```
Please remove this print stmt.

noopurintel commented on tests/github/experimental/workflow/LocalRuntime/testflow_subset_of_collaborators.py:
```diff
@@ -125,6 +125,8 @@ def callable_to_initialize_collaborator_private_attributes(collab_name):
             )
 
     random_ints = random.sample(range(1, len(collaborators) + 1), len(collaborators))
+    print("-------")
```
Please remove these print stmts.

noopurintel commented on tests/end_to_end/workflow/subset_flow.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        List of collaborators ran successfully
+        """
+        log.info("inside join")
```
Please remove this print stmt or move to log.debug.

payalcha replied:
These statements are used in MetaFlow I think. This is how it checks where all it went.

noopurintel replied:
No, there's no such statement in Metaflow - https://github.com/search?q=repo%3ANetflix%2Fmetaflow+%22inside+join%22&type=code

Even if we want to log here, let's print the list of collaborators which ran here

log.info(f"Inside join function with {self.collaborators_ran}")

noopurintel commented on tests/end_to_end/workflow/subset_flow.py:
```diff
@@ -0,0 +1,57 @@
+        """
+        set the collaborator name
+        """
+        log.info(f"Print collaborators {self.name}")
```
Please remove this print stmt.

payalcha replied:
These statements are used in MetaFlow I think. This is how it checks where all it went.

noopurintel replied:
Again, it's not used in Metaflow, but it's okay to have it.

noopurintel commented on tests/end_to_end/workflow/subset_flow.py:
```diff
@@ -0,0 +1,57 @@
+    Testflow to validate working of Subset Collaborators in Federated Flow.
+    """
+
+    def __init__(self, random_ints=[], **kwargs) -> None:
```
Please add a doc-string here covering about random_ints, collaborators, subset_collaborators or any other param involved here.

noopurintel commented on tests/end_to_end/workflow/subset_flow.py:
```diff
@@ -0,0 +1,57 @@
+import logging
```
Copyright section plz.

noopurintel commented on tests/end_to_end/workflow/reference_include_flow.py:
```diff
@@ -0,0 +1,163 @@
+    return matched_ref_dict
+
+
+def validate_references(matched_ref_dict):
```
Please add doc-string. The code logic is not straightaway understood, so provide an example if possible.

payalcha replied:
done

payalcha committed:
Remove extra files

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

noopurintel commented on tests/end_to_end/workflow/reference_include_flow.py:
```diff
@@ -0,0 +1,163 @@
+    return valid_attrs
+
+
+def find_matched_references(collab_attr_list, all_collaborators):
```
Doc-string please.

noopurintel commented on tests/end_to_end/workflow/reference_include_flow.py:
```diff
@@ -0,0 +1,163 @@
+pass
```
Is this `pass` intentional?  Also add copyright section.

noopurintel commented on tests/end_to_end/workflow/reference_include_flow.py:
```diff
@@ -0,0 +1,163 @@
+MIN_COLLECTION_COUNT = 2
+
+
+class Net(nn.Module):
```
Add doc-strings for all the functions present in this file.

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+
+def validate_agg_collab_references(all_collaborators, agg_obj, agg_attrs):
+    """
+    Iterate attributes of aggregator and collaborator to capture the mismatched references.
```
Please add param info as well.

payalcha committed:
Remove extra files

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+
+def validate_agg_attr_ref(agg_attrs, agg_obj):
+    """
+    Verifies aggregator attributes are retained after
```
Please add param info as well.

payalcha replied:
Added

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+
+def validate_collab_references(matched_ref_dict):
+    """
+    Iterate reference list and raise assertion for conflicts
```
Please add param info as well.

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+
+def find_matched_references(collab_attr_list, all_collaborators):
+    """
+    Iterate attributes of collaborator and capture the duplicate reference
```
Please add param info as well.

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+        TestFlowReference.step_two_collab_attrs = []
+
+
+def filter_attrs(attr_list):
```
Doc-string please

noopurintel commented on tests/end_to_end/workflow/reference_flow.py:
```diff
@@ -0,0 +1,297 @@
+        validate_agg_collab_references(inputs, self, agg_attrs)
+
+        all_shared_attr = ""
+        log.info("Reference test summary:")
```
Is this logging needed?

payalcha replied:
These logs are needed if we use Metaflow.

noopurintel commented on .gitignore:
```diff
@@ -15,4 +15,5 @@ venv/*
 .eggs
 eggs/*
 *.pyi
+.metaflow/*
```
Question: is this intentionally added?

payalcha replied:
Yes

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,203 @@
+import logging
```
Copyright section plz.

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,203 @@
+
+log = logging.getLogger(__name__)
+
+def test_exclude_flow(request, fx_local_federated_workflow):
```
Please add doc-string even if one-liner in all the functions.

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,203 @@
+log = logging.getLogger(__name__)
+
+def test_exclude_flow(request, fx_local_federated_workflow):
+    log.info("***** Starting test_exclude_flow *****")
```
Let's avoid using **** or similar characters in our logging. Same goes for all the functions where it is used.

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,203 @@
+    log.info("\n **** Summary of internal flow testing ****\n"
+             "No issues found and below are the tests that ran successfully\n"
+             "1. Number of training completed is equal to training rounds\n"
+             "2. Cli steps and Expected steps are matching\n"
```
Cli -> CLI

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
         workspace_path=workspace_path,
         local_bind_path=local_bind_path,
     )
+
```
2 lines gap.

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+    """
+    from openfl.experimental.workflow.interface import Aggregator, Collaborator
+    from openfl.experimental.workflow.runtime import LocalRuntime
+    collab_callback_func = request.param[0] if hasattr(request, 'param') and request.param else None
```
Keep one line gap. Also adding a comment, on why the import is added inside the function, would be helpful.

payalcha committed:
Add copyright message to workflow files

Add copyright message to the top of each file in the "tests/end_to_end/workflow/" folder.

* Add the copyright message to `exclude_flow.py`
* Add the copyright message to `include_exclude_flow.py`
* Add the copyright message to `include_flow.py`
* Add the copyright message to `internal_loop.py`
* Add the copyright message to `private_attr_both.py`
* Add the copyright message to `private_attr_wo_callable.py`
* Add the copyright message to `private_attributes_flow.py`
* Add the copyright message to `reference_exclude.py`
* Add the copyright message to `reference_flow.py`
* Add the copyright message to `reference_include_flow.py`
* Add the copyright message to `subset_flow.py`

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+    collab_value = request.param[1] if hasattr(request, 'param') and request.param else None
+    agg_callback_func = request.param[2] if hasattr(request, 'param') and request.param else None
+
+    collab_callback_func_name = globals()[collab_callback_func] if collab_callback_func else None
```
Adding a comment that here we change the function name string to a callable function will be helpful for future reference and readability.

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+        local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators_list, backend=backend)
+    local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators_list)
+
+    log.info(f"Local runtime collaborators = {local_runtime.collaborators}")
```
Is this logging required?

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+        collaborators=collaborators_list,
+        runtime=local_runtime,
+    )
+
```
2 lines of gap.

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+        LocalRuntime: An instance of `LocalRuntime` configured with the aggregator,
+                    collaborators, and backend.
+    """
+    from openfl.experimental.workflow.interface import Aggregator, Collaborator
```
same as above comment.

payalcha committed:
Add copyright

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

noopurintel commented on tests/end_to_end/utils/common_fixtures.py:
```diff
@@ -113,3 +121,155 @@ def fx_federation(request):
+    )
+
+
+def init_collaborator_private_attr_index(param):
```
Please move these functions to another file, maybe `wf_helper.py`

payalcha committed:
Review comments fixes

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
docstring

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
docstring

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Doc string

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Add comments

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

payalcha committed:
Review comments

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+
+def test_include_exclude_flow(request, fx_local_federated_workflow):
+    """
+    Test variables which is excluded will not show in next step
```
is -> are
show -> show up

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+def test_include_exclude_flow(request, fx_local_federated_workflow):
+    """
+    Test variables which is excluded will not show in next step
+    Test variables which are included will show in next step
```
show -> show up

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+
+def test_include_flow(request, fx_local_federated_workflow):
+    """
+    Test if variable is included, variables woll show in next step and all other variables will not show
```
Typo -> will* 
show -> show up

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+
+def test_internal_loop(request, fx_local_federated_workflow):
+    """
+    Verify that thru internal loop, rounds to train is set
```
thru -> through

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+@pytest.mark.parametrize("fx_local_federated_workflow", [("init_collaborator_private_attr_index", "int", None )], indirect=True)
+def test_reference_flow(request, fx_local_federated_workflow):
+    """
+    Test reference variables matched through out the flow
```
through out -> throughout

noopurintel commented on tests/end_to_end/test_suites/wf_local_func_tests.py:
```diff
@@ -0,0 +1,249 @@
+
+def test_private_attr_wo_callable(request, fx_local_federated_workflow_prvt_attr):
+    """
+    Set private attribute without callable function i.e thru direct assignment
```
The word 'thru' is informal, so let's use the word 'through' wherever applicable.

payalcha committed:
Review comments

Signed-off-by: Chaurasiya, Payal <payal.chaurasiya@intel.com>

scngupta-dsp commented on .github/workflows/wf_functional_e2e.yml:
```diff
@@ -0,0 +1,95 @@
+---
+#---------------------------------------------------------------------------
+# Workflow to run Task Runner E2E tests via Docker
```
Instead of "Task Runner E2E tests" we can mention "Workflow Interface E2E Tests for LocalRuntime"

scngupta-dsp commented on tests/end_to_end/workflow/exclude_flow.py:
```diff
@@ -0,0 +1,114 @@
+
+log = logging.getLogger(__name__)
+
+class TestFlowExclude(FLSpec):
```
Query: The original test case seems to have a functionality where the class attribute _exclude_error_list_ maintained the step in which the error occurred to help with troubleshooting. Is there a reason why that was removed ?

scngupta-dsp commented on tests/end_to_end/workflow/include_exclude_flow.py:
```diff
@@ -0,0 +1,115 @@
+
+log = logging.getLogger(__name__)
+
+class TestFlowIncludeExclude(FLSpec):
```
Query: Same comment as in TestflowExclude. Wondering the reason to modify the flow to remove _include_exclude_error_list_ which could potentially help with debugging

scngupta-dsp commented on tests/end_to_end/workflow/internal_loop.py:
```diff
@@ -0,0 +1,86 @@
+logging.basicConfig(level=logging.INFO)
+log = logging.getLogger(__name__)
+
+class TestFlowInternalLoop(FLSpec):
```
We can add the purpose of the Flow (missing in original test cases): Testflow to validate the Internal Loop functionality

scngupta-dsp commented on tests/end_to_end/workflow/private_attr_both.py:
```diff
@@ -0,0 +1,142 @@
+log = logging.getLogger(__name__)
+
+class TestFlowPrivateAttributesBoth(FLSpec):
+    """
```
Can we mention that in this test, the private attributes are initialized by both methods: callable as well as dictionary.

scngupta-dsp commented on tests/end_to_end/workflow/private_attr_wo_callable.py:
```diff
@@ -0,0 +1,142 @@
+
+
+class TestFlowPrivateAttributesWoCallable(FLSpec):
+    """
```
Can we mention that in this Test the private attributes are initialized directly

scngupta-dsp commented on tests/end_to_end/workflow/private_attributes_flow.py:
```diff
@@ -0,0 +1,139 @@
+
+class TestFlowPrivateAttributes(FLSpec):
+    """
+    Testflow to validate Aggregator private attributes are not accessible to collaborators
```
Can we mention that in this test, private attributes are initialized by callable

scngupta-dsp commented on tests/end_to_end/workflow/reference_exclude.py:
```diff
@@ -0,0 +1,237 @@
+    """
+
+    step_one_collab_attrs = []
+    step_two_collab_attrs = []
```
What is the reason for removing the functionality related to _all_ref_error_dict_ attribute in the original test case? I am not able to recollect the exact reason this attribute was used and therefore the query

scngupta-dsp reviewed (commented):
To maintain consistency with other documentation I would recommend usage of Workflow Interface and LocalRuntime instead of Workflow APIs and LocalRuntime APIs
